### DISCUSSION TITLE
What is the correct way to config winmerge as merge tool under Windows, and using git-for-windows

### DISCUSSION THREAD

**Discussion by asmwarrior** [WinMerge/winmerge#1853] [General]
Hi, there are many discussion about this topic. I tried several, and I would like the winmerge devs' help.

This is what I'm currently using:

~~~~
[mergetool]
    prompt = false
    keepBackup = false
    keepTemporaries = false

[merge]
    tool = winmerge

[mergetool "winmerge"]
    name = WinMerge
    trustExitCode = true
    cmd = "/c/Program\\ Files/WinMerge/WinMergeU.exe" -u -e -dl \"Local\" -dr \"Remote\" $LOCAL $MERGED $REMOTE
~~~~

When I got a git rebase conflict, I just type the `git mergetool` in the bash shell under windows to fire the winmerge.

Now, the winmerge gives me three panels. The left panel is titled as "Local", and the right panel is titled as "Remote". The middle title don't have a title, it just show the full file path of the currently merged file.

This file contains many `>>>>>>>>` and `<<<<<<<<<` text which is marked as conflict places. Am I correct?

Here are some other things I found in the internet, I'm not sure it is correct.

1, I found in this discussion, some one suggest use something like:

~~~~
cmd = "/c/Program\\ Files\\ \\(x86\\)/WinMerge/WinMergeU.exe" -u -e -dl \"Local\" -dr \"Remote\" $LOCAL $REMOTE $MERGED
~~~~
Note the `$MERGED` is put as the last option.

While, if I use [gitextensions/gitextensions: Git Extensions is a standalone UI tool for managing git repositories. It also integrates with Windows Explorer and Microsoft Visual Studio (2015/2017/2019).](https://github.com/gitextensions/gitextensions), the suggest command is something like:

~~~~
"C:/Program Files/WinMerge/winmergeu.exe" -e -u  -wl -wr -fm -dl "Mine: $LOCAL" -dm "Merged: $BASE" -dr "Theirs: $REMOTE" "$LOCAL" "$BASE" "$REMOTE" -o "$MERGED"
~~~~
But when I used this suggest command, I see the git merge does not shown correctly. Though in the middle panel, their is no text with `<<<<<<<<<` or `>>>>>>>>>>`, but the middle panel is not the automatically merged files, and I guess the middle file are wrong.

2, I found many other resources and discussion here:

[How can I make WinMerge my git mergetool?](https://stackoverflow.com/questions/10713925/how-can-i-make-winmerge-my-git-mergetool), but it has too many answers, and I don't know I would follow which one.

3, I found another discussion in [How to properly use 3 way automatic merge? · WinMerge/winmerge · Discussion #1156](https://github.com/WinMerge/winmerge/discussions/1156), but sadly, it is hard to understand the texts below:

> WinMerge's auto-merge assumes the following order of files, depending on which pane is currently active(It’s a little inconsistent...):
> 
>     Middle pane: assumes that the files are in the order [theirs]-[base]-[mine], and merges the unconflicted changes in left (theirs) and right (mine) into middle (base).
>     Right pane: assumes that the files are in the order [base]-[theirs]-[mine], and merges the non-conflicting changes in middle(theirs) into right(mine).
>     Left pane: assumes that the files are in the order [mine]-[theirs]-[base], and merges non-conflicting changes in the middle (theirs) into the left (mine).

Can you explain it?

4, when I looked at the winmerge command line option document page: [Command line - WinMerge 2.16 Manual](https://manual.winmerge.org/en/Command_line.html), some options are marked as TBD, I'm not sure why they are not defined?
Also, there is a section named: [Resolving conflict files](https://manual.winmerge.org/en/Compare_files.html#ResolveConflictFiles), can this option be used in git's merge tool config setting?

Thanks.

**Comment by sdottaka** [comment]
First of all, I do not recommend WinMerge as a 3-way merge tool, as it is not superior to other tools.
Sorry for not answering all your questions, but if you want to automatically merge in the middle pane when WinMerge starts, you can specify the `-am` command line option.

I haven't tried it yet, but I believe it would look something like this.
~~~
"C:/Program Files/WinMerge/winmergeu.exe" -e -u -wl -wr -am -fm -dl "Mine: $LOCAL" -dm "Merged: $BASE" -dr "Theirs: $REMOTE" "$LOCAL" "$BASE" "$ REMOTE" -o "$MERGED"
~~~

When merged automatically by WinMerge, conflicting locations are not marked as <<<< >>>>. The location pane shows the conflicts in red." Next Conflict" icon to go to the conflicted location.

**Comment by asmwarrior** [comment]
Hi, @sdottaka thanks for the reply.

> First of all, I do not recommend WinMerge as a 3-way merge tool, as it is not superior to other tools.

But I just see that winmerge is a tiny tool(only 20M+ bytes) which can do the job. I have tried kdiff3 under Windows, it has 300M+ bytes in my local disk, I'm not sure why a diff/merge tool need so much disk space.

I just tried the `-am` option, it works. But I just have another question:

~~~~
"C:/Program Files/WinMerge/winmergeu.exe" -e -u -wl -wr -am -fm -dl "Mine: $LOCAL" -dm "Merged: $BASE" -dr "Theirs: $REMOTE" "$LOCAL" "$BASE" "$REMOTE" -o "$MERGED"
~~~~

The option here `-o "$MERGED"` is that when we close the winmerge, the saved middle panel file will overwrite the final merged file(which is used to commit). Since git already prepare the `$MERGED` file, why does winmerge did it again in the `"$LOCAL" "$BASE" "$ REMOTE"`? That's why I suggest using `"$LOCAL" "$MERGED" "$REMOTE"` instead.

There is one issue that after the merge, winmerge dose not give notification if it still has conflicts in the final file. Some guy in stackoverflow gives an workaround, see this post: https://stackoverflow.com/a/30699239/154911 with Vonc's last comment. So, can this be a feature request for winmerge?

Thanks.

**Reply by sdottaka** [reply]
> why does winmerge did it again in the "$LOCAL" "$BASE" "$REMOTE"? That's why I suggest using "$LOCAL" "$MERGED" "$REMOTE" instead.

The $MERGED file probably has a conflict marker like <<<< >>>>, so if you don't mind it, you can specify $MERGED in the middle. If you specify $MERGED, $MERGE is already an auto-merged file, so you don't need the `-am` option.

>There is one issue after the merge, winmerge does not give notification if it still has conflicts in the final file. 

You're right, it would be nice to have a notification if WinMerge was closed while there were conflicts.

**Comment by asmwarrior** [comment]
Thanks for the explanation. Yes, the `$MERGED` is a file which is already auto-merged by git. For me, I'd prefer the `-am` option. One question is that does the "auto-merged by git" is the same as "auto-merge by winmerge"? If they are the same, I certainly prefer using the `-am` option, because removing the `<<<<<<<>>>>>>>` like markers is cumbersome. 

I wrote a feature request here: [It would be nice to notify the user if there are un-resolved conflicts in the merged file before winmerge closed · Issue #1855 · WinMerge/winmerge](https://github.com/WinMerge/winmerge/issues/1855)

**Reply by sdottaka** [reply]
>One question is that does the "auto-merged by git" is the same as "auto-merge by winmerge"?

I think they do something similar, but git might do smarter merging.

**Reply by asmwarrior** [reply]
OK, thanks.  We have updated this stackoverflow answer to mention your suggestion, also the feature request here: https://stackoverflow.com/a/30699239/154911

**Comment by Chealer** [comment]
Hi @sdottaka ,

> First of all, I do not recommend WinMerge as a 3-way merge tool, as it is not superior to other tools.

What do you mean by that? Do you mean you *recommend not to* use WinMerge as a 3-way merge tool as it would be *inferior* to other tools? And if so, which other tools would be superior to WinMerge?

**Reply by sdottaka** [reply]
Actually, I wanted WinMerge to be able to merge in 4 panes like kdiff3, but it still only has 3 panes.
Free tools that are better 3-way merge tools than WinMerge include kdiff3, p4merge, and tortoisemerge.

**Reply by Chealer** [reply]
Thank you @sdottaka 
P4Merge is proprietary, but do you have a recommendation between KDiff3 and TortoiseMerge? I haven't used TortoiseMerge without TortoiseSVN yet.

Note that version 2.5 (2015) of Git [has added builtin integration for merging with Winmerge](https://github.com/git/git/commit/3e4f2373b295e19dbc1a6bd0e743f65b500b7bda), which seems to be a bad idea if Winmerge does so poorly for that task.

Similarly, TortoiseHg 6.5.1 has builtin [integration for merging](https://tortoisehg.readthedocs.io/en/latest/common.html#adding-tools) with Winmerge, as can be seen in `defaultrc\MergeTools.rc`:
```
winmergeu.args=/e /x /u /wl /wr /dl local /dr other $local $output $other
winmergeu.checkchanged=True
winmergeu.diff3args=/r /e /x /u /wl /wr /dl '$plabel1' /dm '$clabel' /dr '$plabel2' $parent1 $child $parent2
winmergeu.diffargs=/r /e /x /u /wl /dl '$plabel1' /dr '$clabel' $parent $child
winmergeu.dir3diff=True
winmergeu.dirdiff=True
winmergeu.fixeol=True
winmergeu.gui=True
winmergeu.priority=-10
winmergeu.regkey=Software\Thingamahoochie\WinMerge\
winmergeu.regkeyalt=Software\Wow6432Node\Thingamahoochie\WinMerge\
winmergeu.regname=Executable
```

**Reply by Chealer** [reply]
TortoiseMerge is ([basically](https://stackoverflow.com/questions/25196006/install-just-tortoisemerge-software-and-context-menu-items)) not available as a standalone application, so is a doubtful choice except for Subversion users. Meld (GNOME) worked for me but seemed really poor for merges (incomplete and buggy).

So I went with KDiff3, and after forcing it to use the proper encoding (instead of the wrong one it chose by default for some reason), I managed to get the job done with help from a colleague. It's far from being as intuitive as WinMerge normally is, but when you spend enough time studying it, it seems to work **reasonably** well.

**Reply by asmwarrior** [reply]
Hi, using TortoiseGitMerge standalone version is OK on my Windows, see my github repo here: [asmwarrior/TortoiseGitMerge_Standalone: Standalone versions of TortoiseGitMerge - TortoiseGit doesn't offer it on its own.](https://github.com/asmwarrior/TortoiseGitMerge_Standalone), I just extract one exe file and 3 dlls from the TortoiseGit installation.

Hope that helps.
